### PR TITLE
[Wysiwyg] Upgrade CKEditor4 to CKEditor5

### LINKED ISSUE #276: [Feature Request] A new WYSIWYG editor
Implement a modern WYSIWYG editor with markdown output.
We could go with raw HTML but we'd need to make Twig render it right away instead of sanitizing it into raw text.

One possible solution would be [CKEditor 5](https://ckeditor.com/). It has [markdown output](https://ckeditor.com/docs/ckeditor5/latest/features/markdown.html) support.

### PR DESCRIPTION
**Pull request by shmshd**
Upgrade CKEditor4 to CKEditor5 with Gulp task.

Closes #276

### CODE REVIEW

**Review by jaapmarcus — CHANGES_REQUESTED** [review]
There seems to be something wrong with the texture heights:

Screenshot 2023-01-22 at 11 02 09
Screenshot 2023-01-22 at 11 01 30

Also for security reasons we currently only support md format making CKEditor useless

Screenshot 2023-01-22 at 11 04 55

**Comment by evrifaessa** [comment]
> Also for security reasons we currently only support md format making CKEditor useless

@jaapmarcus - We already are using CKEditor 4. This PR upgrades it to CKEditor 5 and CKEditor 5 can be tweaked to output Markdown.

This is a nice PR and imo would be good to merge with a few adjustments: https://ckeditor.com/docs/ckeditor5/latest/features/markdown.html

Also thanks for integrating it as a Gulp task. Much easier to maintain for the long run.

**Comment by BelleNottelling** [comment]
> > Also for security reasons we currently only support md format making CKEditor useless
> 
> @jaapmarcus - We already are using CKEditor 4. This PR upgrades it to CKEditor 5 and CKEditor 5 can be tweaked to output Markdown.
> 
> This is a nice PR and imo would be good to merge with a few adjustments: https://ckeditor.com/docs/ckeditor5/latest/features/markdown.html
> 
> Also thanks for integrating it as a Gulp task. Much easier to maintain for the long run.

Yeah, it'll be a great upgrade over either of the WYSIWYG editors, once it's configured to output markdown

**Comment by shmshd** [comment]
In this case, we will have to build the CKEditor ourself instead of using their basic build. Building ourself needs Webpack and their quite some dev packages. My concern is, should we separate the packages to be used from the module itself as in having `package.json` for Wysiwyg in its own directory? This way main project packages are separated and this will contain only packages for Wysiwyg editor(s). If needed we can trigger npm installation and build from main Gulp task build command which can save some time for developers/contributors.

**Comment by evrifaessa** [comment]
@shmshd Sounds like a good plan :)

**Comment by shmshd** [comment]
Test fail is due to missing `node_modules ` in Wysiwyg dir. We need to trigger package installation from the main Gulpfile. Test the current commits and let me know.

**Comment by evrifaessa** [comment]
We will do our best to test this thoroughly. It likely won't be included with the next release, but that's for the best to avoid possible issues.

**Comment by shmshd** [comment]
> to the lock file we should commit that as well to the lock file

oops. missed it.

**Comment by evrifaessa** [comment]
On pages that are supposed to display more than one instance of CKEditor, only the first one loads.

image

**Comment by evrifaessa** [comment]
While doing changes in the CKEditor content, the value of the underlying textarea won't be updated.

Try doing some changes in "/admin/extension/settings/system" and saving the settings. After you reload, you will notice the value wasn't actually updated.

[Screencast from 2023-01-23 16-53-56.webm](https://user-images.githubusercontent.com/35808275/214057335-d4d04df7-4e16-4e47-b6c4-5535c9f608c6.webm)

**Comment by evrifaessa** [comment]
> I wonder if our api call get priority over ckeditor events when you press update settings

Check this: https://ckeditor.com/docs/ckeditor5/latest/installation/getting-started/getting-and-setting-data.html#manually-retrieving-the-data

Because we are using AJAX/fetch to do the requests instead of using regular HTML forms, we need to explicitly call `getData()` to get the editor content.

**Comment by evrifaessa** [comment]
I'll see if I can resolve the two issues I've listed above

**Comment by evrifaessa** [comment]
Pushed. Multiple editors in a page should now work. The form handler also now uses the correct content.

**Comment by evrifaessa** [comment]
image

**Comment by evrifaessa** [comment]
A thing is, tables won't work. CKEditor outputs them in raw HTML.

The CKEditor documentation says this:

> Please bear in mind that the Markdown data processor does not support all rich text features. The [Markdown syntax](https://daringfireball.net/projects/markdown/syntax) is very simple and only supports limited formatting options.
This means that advanced formatting like list styles, table styles or page break markers will be stripped in the effecting data. These are not supported by Markdown and therefore cannot be converted from HTML to Markdown.

Even if it did output tables in the [GitHub Flavored Markdown](https://github.github.com/gfm/#tables-extension-) format, our Markdown to HTML converter that we use in the backend doesn't support the syntax either.

My opinion would be to remove the table functionality.

image

**Comment by BelleNottelling** [comment]
Commonmark supports GitHub flavored markdown, I specifically enabled it. 

Tables will work

**Comment by BelleNottelling** [comment]
> > Commonmark supports GitHub flavored markdown, I specifically enabled it.
> > 
> > Tables will work
> 
> I've manually edited the source and added tables with the GFM syntax. Not sure if this is how they are supposed to look in the Huraga frontend.
> 
> image
> image
> image
> 

Seems like it, it's clearly formatting it as a table. You can check in the page source to see what it's being outputted as. 

My other project uses the same markdown parser + two extensions, so all of markdown syntax seen on [this page](https://antcms.org/docs/features) except emojis and embedded content will be supported in FOSSBilling

**Comment by evrifaessa** [comment]
Yup, it will correctly display it as a `<table>`.

However, CKEditor's own table implementation won't use the GFM table syntax and what I did was editing the content source to manually add the table.

**Comment by shmshd** [comment]
> However, CKEditor's own table implementation won't use the GFM table syntax and what I did was editing the content source to manually add the table.

The issue is explained here https://github.com/ckeditor/ckeditor5/issues/12793#issuecomment-1375533358
TL;DR Tables must have header as per the GFM spec. If header is not specified, it goes to HTML way.

Better solution in my opinion is to enforce the header as mentioned in here https://github.com/ckeditor/ckeditor5/issues/12793#issuecomment-1375549722

Did not test it myself though.

**Comment by evrifaessa** [comment]
Can we add image functionality? The backend can successfully understand and render images.
https://ckeditor.com/docs/ckeditor5/latest/features/images/images-overview.html

It doesn't necessarily need to be able to handle files uploaded from local disks. Just allowing images to be embedded using external links should be fine.

image

**Comment by jaapmarcus** [comment]
```
CKEditor
                .create(element, {table: { defaultHeadings: { rows: 1 } }})
```

Works fine

**Comment by shmshd** [comment]
> In fact it currently seems like images are completely blocked. CKEditor will remove the images even if you edit the source and manually add the image. To test if it would work, I edited the setting manually by connecting to the database directly.

We need an additional package/plugin to be added for images.

**Comment by shmshd** [comment]
Insert image does have both local file upload and add via URL options, but the local file upload requires a [custom upload adapter](https://ckeditor.com/docs/ckeditor5/latest/framework/guides/deep-dive/upload-adapter.html). Due to this, we will now allow image insert via URL only, until someone comes up with an adapter.

**Comment by evrifaessa** [comment]
I don't think the backend currently is able to handle uploaded images. 

It makes sense to only allow inserting images with URLs for now.

**Comment by shmshd** [comment]
Direct image URL paste to the editor is also enabled with `autoimage ` plugin.

Is there any other issue to be addressed?

**Comment by evrifaessa** [comment]
https://github.com/FOSSBilling/FOSSBilling/pull/713#issuecomment-1400485593

These three plugins were not loaded.

**Comment by evrifaessa** [comment]
> The minimum required `npm` version will be `v7` due to the `workspace` usage. Should we mention it in `package.json`?

I’m okay with that. NPM 9 is out now and it’s been 2 years since NPM 7 was made generally available.

**Review by jaapmarcus — APPROVED** [review]
Changes look fine and next step is dropping support for the markdownit!

**Comment by shmshd** [comment]
@jaapmarcus could you check my comment on your CSS change?

**Comment by jaapmarcus** [comment]
> @jaapmarcus could you check my comment on your CSS change?

What comment?

**Comment by shmshd** [comment]
> What comment?

image

**Comment by jaapmarcus** [comment]
Comments are set as pending so not pushed. 

min-height is fine for me :)

**Comment by evrifaessa** [comment]
It seems mostly good to go. I want to test it more thoroughly today.

To avoid possible issues, we will not include it with the next release and wait for the one after that.

**Comment by evrifaessa** [comment]
For some reason, the "New Custom Page" form (/admin/custompages) will not work with CKEditor 5 enabled.

I know the module is old and the interface is heavily outdated, but we should sort this to make sure it still functions regularly.

**Comment by John-S4** [comment]
Custom pages has some issues apart from that. It will return an error if you try to edit a page even before this PR.

**Comment by evrifaessa** [comment]
> Custom pages has some issues apart from that. It will return an error if you try to edit a page even before this PR.

Still, this particular one seems to be caused by our implementation of CKEditor. Can confirm that it works on the latest release (0.2.9) but not on this branch.

**Comment by jaapmarcus** [comment]
> > Custom pages has some issues apart from that. It will return an error if you try to edit a page even before this PR.
> 
> Still, this particular one seems to be caused by our implementation of CKEditor. Can confirm that it works on the latest release (0.2.9) but not on this branch.

It also still works with mark-down editor.. and the main branch

**Comment by shmshd** [comment]
> For some reason, the "New Custom Page" form (`/admin/custompages`) will not work with CKEditor 5 enabled.

The issue is because of the `required` attribute in the `<textarea>`. Element with required should be focusable. CKEditor always set `display: none;` style to the original element. When using `required` on the hidden element, browsers get triggered. Removing `required` attribute simply fixes the issue.

**Comment by evrifaessa** [comment]
> The issue is because of the `required` attribute in the `<textarea>`. Element with required should be focusable. CKEditor always set `display: none;` style to the original element. When using `required` on the hidden element, browsers get triggered. Removing `required` attribute simply fixes the issue.

We need to address that. Some fields need to be kept as required and we can't just go and remove the `required` attribute.

**Comment by shmshd** [comment]
> We need to address that. Some fields need to be kept as required and we can't just go and remove the `required` attribute.

Then my suggestion would be to remove that attribute and set validation after the initialization of CKEditor.
```js
if (element.hasAttribute('required')) {
  element.removeAttribute('required');
  element.reportValidity();
}
```
Does anyone have a better approach?

**Comment by BelleNottelling** [comment]
> > We need to address that. Some fields need to be kept as required and we can't just go and remove the `required` attribute.
> 
> Then my suggestion would be to remove that attribute and set validation after the initialization of CKEditor.
> 
> ```js
> if (element.hasAttribute('required')) {
>   element.removeAttribute('required');
>   element.reportValidity();
> }
> ```
> 
> Does anyone have a better approach?

Fine with me as long as we  can make it only target the WYSIWYG elements when using CKEditor

**Comment by evrifaessa** [comment]
@shmshd I've pushed 2 commits for the required fields. Feel free to revert them if you have a better idea.

**Comment by shmshd** [comment]
@evrifaessa I thought we were not supposed to remove (or replace) the `required` attribute from forms since it is an issue with CKEditor5. Changing the behaviour of the forms for an editor seems a bit off as it affects other editor(s). Just my thought.

**Comment by evrifaessa** [comment]
Oh you're absolutely right. I forgot about other editors.

Feel free to revert the two commit, thanks for your contribution.

**Comment by jaapmarcus** [comment]
> @evrifaessa I thought we were not supposed to remove (or replace) the `required` attribute from forms since it is an issue with CKEditor5. Changing the behaviour of the forms for an editor seems a bit off as it affects other editor(s). Just my thought.

We should drop the other editor support as well. It doesn't make sense to include 2 editors that can do the same. But one is more restricted...

**Comment by evrifaessa** [comment]
> We should drop the other editor support as well. It doesn't make sense to include 2 editors that can do the same. But one is more restricted...

Yes, but for future-proofing just in case we introduce another editor in the future, this makes sense.

**Comment by shmshd** [comment]
> Yes, but for future-proofing just in case we introduce another editor in the future, this makes sense.

This is the reason I added much more flexibility to this module. In the future, as an alternative we could add Froala (which is highly customizable) editor or some other popular wysiwyg editor. That needs a discussion of its own.

**Comment by evrifaessa** [comment]
@shmshd - I'm afraid this doesn't work.

When tested on `/admin/custompages`, it will remove the "required" attribute but won't actually check for anything. It will straight up pass it to the API wrapper and proceed with the request.

Maybe what you are looking for is in `/themes/admin_default/assets/js/fossbilling.js`. We prevent the default `submit` event and handle doing the request ourselves.

I know the code there is messy, but the part you specifically should be looking for is the `apiForm` function inside the `bb` object.

**Comment by shmshd** [comment]
> @shmshd - I'm afraid this doesn't work.

It should be fine now

**Review by evrifaessa — APPROVED** [review]
Everything looks good to me. I'd definitely want the other team members to test before we merge it though. Thanks for the very valuable contribution, we are looking forward to more :)

**Comment by shmshd** [comment]
@evrifaessa Just for your info, your editor seems to not respect `.editorconfig`. `package.json` file ending line is removed (https://github.com/FOSSBilling/FOSSBilling/blob/main/.editorconfig#L16).

**Comment by evrifaessa** [comment]
> @evrifaessa Just for your info, your editor seems to not respect `.editorconfig`. `package.json` file ending line is removed ([`main`/.editorconfig#L16](https://github.com/FOSSBilling/FOSSBilling/blob/main/.editorconfig?rgh-link-date=2023-02-08T10%3A38%3A48Z#L16)).

I'll look into that. Thanks.

**Review by BelleNottelling — APPROVED** [review]
Works well, it'll be great to have this in place of the existing WYSIWYG editor.
Thank you for the quality contribution!

**Review by jaapmarcus — APPROVED** [review]
